### ISSUE TITLE
AutoLogin and Displaymanager module using #defaultDesktopEnvironment in conf on 14 different desktops

### ISSUE THREAD

**Issue by arcolinuxz** [calamares/calamares#1205]
Goal : autologin into all desktops after installation calamares and rebooting

Calamares 3.2.10 from this github.
Lightdm only displaymanager 1.30.0

ArcoLinux has 14 desktops.

To maintain and fast development we use XFCE as tool to deliver all desktops to our users.
This means you will always have XFCE running on a livedvd. If we let the module decide to choose its autologin desktop environment, it will always choose XFCE.

Recently I discovered this possibility.

```
#Enable the following settings to force a desktop environment in your displaymanager configuration file:
#defaultDesktopEnvironment:
executable: "startkde"
desktopFile: "plasma"
```
That sounded promising and all calamares configs have been changed to include the file displaymanager.conf and the content of the intended desktop environment after installation.

Example Plasma

`# Configure one or more display managers (e.g. SDDM)
 with a "best effort" approach.
---
#The DM module attempts to set up all the DMs found in this list, in that precise order.
#It also sets up autologin, if the feature is enabled in globalstorage.
#The displaymanagers list can also be set in globalstorage, and in that case it overrides anything set up here.
displaymanagers:

  - lightdm

#Enable the following settings to force a desktop environment in your displaymanager configuration file:
#defaultDesktopEnvironment:
executable: "startkde"
desktopFile: "plasma"

#If true, try to ensure that the user, group, /var directory etc. for the
#display manager are set up correctly. This is normally done by the distribution
#packages, and best left to them. Therefore, it is disabled by default.
basicSetup: false

#If true, setup autologin for openSUSE. This only makes sense on openSUSE
#derivatives or other systems where /etc/sysconfig/displaymanager exists.
sysconfigSetup: false`

-----------------------------------------------------------------

4 desktops from the 14 we have, will boot into its desktop without manual login after installation and rebooting.

**Maybe I am doing something wrong on my part. But 10 desktops will not autologin after installation and rebooting.**

Here is a summary of all my settings - all displaymanager.conf are the same except from lines 13 and 14 - the executable and the desktop file.

- awesome
executable: "awesome"
desktopFile: "awesome"
- bspwm
executable: "bspwm"
desktopFile: "bspwm"
- budgie
executable: "budgie-desktop"
desktopFile: "budgie-desktop"
- cinnamon
executable: "cinnamon-session-cinnamon"
desktopFile: "cinnamon"
- deepin
executable: "startdde"
desktopFile: "deepin"
- gnome
executable: "gnome-session"
desktopFile: "gnome"
- i3
executable: "i3"
desktopFile: "i3"
- lxqt
executable: "startlxqt"
desktopFile: "lxqt"
- mate
executable: "mate-session"
desktopFile: "mate"
- openbox
executable: "openbox-session"
desktopFile: "openbox"
- plasma
executable: "startkde"
desktopFile: "plasma"
- qtile
executable: "qtile"
desktopFile: "qtile"
- xfce
standard we have xfce in there already
- xmonad
executable: "xmonad"
desktopFile: "xmonad"

These 4 will autologin into the desktop after installation and reboot:
- gnome
- budgie
- xfce
- plasma

The others will not and I need another pair of eyes to shed some light as to why not?

These 4 desktops will change the content of /etc/lightdm/lightdm.conf and set
 
`user-session=xfce` to the desired desktop like plasma, gnome, budgie,...

Calamares will not change the lightdm.conf file to the corresponding desktops for the other 10.

Here is the standard lightdm.conf we start with on all iso's.
https://github.com/arcolinux/arcolinux-iso/blob/master/archiso/airootfs/etc/lightdm/lightdm.conf

**Comment by adriaandegroot** [comment]
That's really weird, since nothing in the code checks for anything: it just runs `sed` on the `lightdm.conf` to substitute in the DE name. I'll try to get one of the non-working Arcolinux ISO's locally and add a bunch of debugging to the module.

**Comment by arcolinuxz** [comment]
How can I assist in this... 
I see the same issue in 3.2.15 version.

**Comment by arcolinuxz** [comment]
Every desktop has a displaymanager.conf. Here example from qtile.

https://github.com/arcolinux/arcolinuxb-calamares-qtile/blob/master/calamares/modules/displaymanager.conf

**Comment by kkofler** [comment]
Why is `#defaultDesktopEnvironment:` commented out?

**Comment by arcolinuxz** [comment]
you mean to say if we delete the # it may work.

trying immediately ...

xfce, budgie, gnome and plasma do work however with the hashtag

https://github.com/arcolinux/arcolinuxb-calamares-plasma/blob/master/calamares/modules/displaymanager.conf

https://github.com/arcolinux/arcolinuxb-calamares-gnome/blob/master/calamares/modules/displaymanager.conf

I seem to hear I need to delete the hashtag ...

**Comment by arcolinuxz** [comment]
I have just tried it with Awesome desktop.
# is gone
![calamares-error](https://user-images.githubusercontent.com/36679406/67164614-b2a28380-f37c-11e9-901d-f3b152b7d9cd.png)

That is what you get then.

**Comment by abucodonosor** [comment]
@arcolinuxz 

Your configuration has broken formating.

**Comment by arcolinuxz** [comment]
clearly

**Comment by arcolinuxz** [comment]
Keep track of all the Desktops and autologin (I will edit this topic every time I build a new iso)
Calamares 3.2.15
ArcoLinux 19.12.2


- ArcoLinuxB Mate - no autologin

**Comment by arcolinuxz** [comment]
Can someone point me to the actual coding where the change takes place as described above?
@adriaandegroot 

> That's really weird, since nothing in the code checks for anything: it just runs `sed` on the `lightdm.conf` to substitute in the DE name. I'll try to get one of the non-working Arcolinux ISO's locally and add a bunch of debugging to the module.

https://arcolinux.com/how-to-theme-archmerge-xfce-any-way-you-want/

**Comment by adriaandegroot** [comment]
You mention two ISO builds with different results, above:
 - Mate (no autologin)
 - Budgie (autologin)

I'm going to assume that in both cases, you have the same **display** manager, with different **desktop** environments to (auto)login into. Can you post the two `lightdm.conf` files that are used then? Because they must be different in some way.

In `main.py` (of the *displaymanager* module) around line 168, there is a list of executables and desktop-environment `.desktop` files. Those are used to guess which environments are installed and how to start them. The actual work of configuring *lightDM* is done by `class DMlightdm(DisplayManager):` around line 612, with `set_autologin()` the most important part.

It should already be logging enough to the session log to be able to see why it is (or isn't) enabling autologin.

**Comment by arcolinuxz** [comment]
Dag Adriaan,

Thanks to your input I figured out a way to autologin on ArcoLinux iso's (35) and 15 desktops.

1. we changed our config so that displaymanager comes AFTER packages module. After your text I know now that IF you set the desktop explicitly (and we do but calamares) it should recognize it and replace it in the lightdm.conf.. But whatever we feed Calamares in displaymanager.conf /usr/bin/i3 or i3 etc... calamares will NOT use it.
Instead calamares will take the first desktop it comes across in the listing at the top in main.py of displaymanager. And that suddenly made sense. Budgie, Gnome and Xfce were never a problem. That is because of the order you put them in at the top.

First action is change the order when displaymanager is going to check what is inside /usr/share/xsessions. Display manager module should run AFTER packages module then we have XFCE deleted when we install Xmonad for example.

We use Xfce to deliver 15 desktops. So xfce was present in /usr/share/xsession when displaymanager went to look. Changing the sequence means xfce is then deleted.

https://github.com/arcolinux/arco-calamares-config/blob/master/calamares/settings.conf.


2. second thing we had to do is work with the variable not being applied from the displaymanager.conf.
We have an iso with 3 desktops that should start with xfce and an iso with 7 desktops that should start with openbox. Changing the order fixed the autlogin.

https://github.com/arcolinux/arcolinux-pkgbuild-3party/blob/master/calamares/arco-calamares-3.2.16-29/dm_main.py

For now we have what we want. 
All 15 desktop have autologin.

The question remains why we can not set it via the parameter in displaymanager.conf.

Any questions ... ask away.

It seems whatever I type in in the displaymananager.conf calamares will not take a look at it not if it says /usr/bin/i3 or i3.

**Comment by arcolinuxz** [comment]
This is how we make calamares work for 15 desktops and get an automatic login.

```
desktop_environments = [
    DesktopEnvironment('/usr/bin/startplasma-x11', 'plasma'),  # KDE Plasma 5.17+
    DesktopEnvironment('/usr/bin/startkde', 'plasma'),  # KDE Plasma 5
    DesktopEnvironment('/usr/bin/startkde', 'kde-plasma'),  # KDE Plasma 4
    DesktopEnvironment(
        '/usr/bin/budgie-desktop', 'budgie-desktop'  # Budgie v10
        ),
    DesktopEnvironment(
        '/usr/bin/budgie-session', 'budgie-desktop'  # Budgie v8
        ),
    DesktopEnvironment('/usr/bin/gnome-session', 'gnome'),
    DesktopEnvironment('/usr/bin/cinnamon-session-cinnamon', 'cinnamon'),
    DesktopEnvironment('/usr/bin/mate-session', 'mate'),
    DesktopEnvironment('/usr/bin/enlightenment_start', 'enlightenment'),
    DesktopEnvironment('/usr/bin/lxsession', 'LXDE'),
    DesktopEnvironment('/usr/bin/startlxde', 'LXDE'),
    DesktopEnvironment('/usr/bin/lxqt-session', 'lxqt'),
    DesktopEnvironment('/usr/bin/pekwm', 'pekwm'),
    DesktopEnvironment('/usr/bin/pantheon-session', 'pantheon'),
    DesktopEnvironment('/usr/bin/startdde', 'deepin'),
    DesktopEnvironment('/usr/bin/startxfce4', 'xfce'),
    DesktopEnvironment('/usr/bin/openbox-session', 'openbox'),
    DesktopEnvironment('/usr/bin/i3', 'i3'),
    DesktopEnvironment('/usr/bin/awesome', 'awesome'),
    DesktopEnvironment('/usr/bin/bspwm', 'bspwm'),
    DesktopEnvironment('/usr/bin/herbstluftwm', 'herbstluftwm'),
    DesktopEnvironment('/usr/bin/qtile', 'qtile'),
    DesktopEnvironment('/usr/bin/xmonad', 'xmonad'),
    DesktopEnvironment('/usr/bin/dwm', 'dmw'),
    DesktopEnvironment('/usr/bin/jwm', 'jwm'),
]
```

**Comment by dj--alex** [comment]
Mate desktop  autologin still no work. :(( 
Install latest available Calamares from Fresh mint 19 install.
Creates image ISO more than 5 times

users.conf settings IS ignored
---
doAutologin: true
setRootPassword: false
sudoersGroup:    sudo
defaultGroups:

**Comment by adriaandegroot** [comment]
Without a PR (with the changes Erik applied) or a **specific** scenario to test (along the lines of "Install this ISO [link], notice the target system has display manager *X* and environment *Y*") there's not much that is going to happen in this ticket.

Linux Mint 19 (as in, the thing from linuxmint.com) doesn't use Calamares for installation, so I'm not sure where that is supposed to come in.

**Comment by arcolinuxz** [comment]
rediscovered this post 
I will make a PR

**Comment by adriaandegroot** [comment]
Presumed fixed with #1658 (if not, @dj--alex in particular, please open a **new** issue with specific steps and links to a relevant ISO). Thanks @arcolinuxz for the PR.
